### ISSUE TITLE
Allow property editors to bind complex objects to the URL

### ISSUE THREAD

**Issue by scothis** [rackt/react-router#1365]
It's sometimes necessary to bind values more complex than strings to the URL. I'd like to see a facility where a property editor can be defined for a param or query property that is able to convert more complex objects.

A potential API could add an `editors` prop to the `Route` component. The prop is a object whose keys represent the params in the route, and value is an object with `stringify` and `parse` methods.

A simple case is mapping to an array of strings, and could look like:

```jsx
<Route
  name='layers-edit'
  path='/layers/:layerIds'
  handler={LayersEdit}
  editors={{
    layerIds: {
      parse: (layerIds) => layerIds.split(','),
      stringify: (layerIds) => layerIds.join(',')
    }
  }} />
```

The real value of editors is that creating `Link`s to these routes is much simpler. Each components that creates a `Link` should not need to be aware of the serialization format used within the URL.

In general, I think these methods should be synchronous, but supporting async editors could open up even more interesting possibilities.

If you like this idea, I'm willing to take a pass at implementing.

**Comment by ryanflorence** [comment]
You can probably implement with your own Link

**Comment by scothis** [comment]
> You can probably implement with your own Link

This is basically what I'm doing right now. It's passible, but could be much cleaner.

What I don't like is:
- other components need to know about and use the custom Link
- each route handler (and children) need to parse the prop

**Comment by mjackson** [comment]
I don't like the idea of storing potentially complex data types directly in the pathname portion of the URL. If you need to associate a complex object with a  certain location, you already have two options (this is assuming the 1.0 API, but it looks like you're using 0.13.3):

1. Put your object in the query string, which already has well-defined mechanisms for dealing with complex data types (though it assumes every value is just a simple string) or
2. Put your object in `location.state` using `<Link ... state={{ stuff: 'here' }}/>` (only available in 1.0)

In the second option your location state will be serialized using JSON so you can include anything JSON supports.
